### ISSUE TITLE
sync_files errors

### ISSUE THREAD

**Issue by dikim33** [oscar-cluster/oscar#565]
**Reported by naughtont on 30 Jul 2009 21:30 UTC**
sync-files prints lots of warning and error messages and doesn't actually transfer/update the files on the nodes.  See below for example output.

sync-files Version: 2.5.2-1
Distro: Ubuntu 8.04.3 (LTS) on x86
OSCAR Version: 6.0.4 (devel)
System: XTORC

```
 #
 # 1) Create new user
 #
 # 2) Confirm account files are set to sync in '/opt/sync_files
 #    /etc/sync_files.conf'
 #
 # 3) Run sync_files
 #

node0:$ sudo /opt/sync_files/bin/sync_files 
--> Now parsing command-line options
--> Now defining variables from config file
--> Ignore weird lines in the next Setting x to y section
--> Setting syncfile to ARRAY(0x899077c)
--> Setting cronmodify to 1
--> Setting getent to 1
--> Setting usec3 to 1
--> Setting scphost to ARRAY(0x89906b0)

=============================================================================
== [#
 # 4) Check for new user on a given node of cluster
 #

node0:$ ssh node1 grep <NewUsername> /etc/passwd
node0:$

 #
 # 5) Try running sync_files with '--force' option...
 #

node0:$ sudo /opt/sync_files/bin/sync_files --force
--> Now parsing command-line options
--> Now defining variables from config file
--> Ignore weird lines in the next Setting x to y section
--> Setting syncfile to ARRAY(0x899068c)
--> Setting cronmodify to 1
--> Setting getent to 1
--> Setting usec3 to 1
--> Setting scphost to ARRAY(0x89905c0)

=============================================================================
== [sync_file](sync_file]
=============================================================================

)
=============================================================================

Use of uninitialized value in numeric eq (==) at /opt/sync_files/bin/sync_files line 791.
Use of uninitialized value in numeric eq (==) at /opt/sync_files/bin/sync_files line 603.
image:
$VAR1 = 'oscarimage';
---------------
Subcluster defined by image oscarimage wasn't validated!
... skipping it ...
Use of uninitialized value in numeric eq (==) at /opt/sync_files/bin/sync_files line 637.
Use of uninitialized value in numeric eq (==) at /opt/sync_files/bin/sync_files line 603.
image:
$VAR1 = 'oscarimage';
---------------
Subcluster defined by image oscarimage wasn't validated!
... skipping it ...
Use of uninitialized value in numeric eq (==) at /opt/sync_files/bin/sync_files line 637.
Use of uninitialized value in numeric eq (==) at /opt/sync_files/bin/sync_files line 603.
[File /opt/sync_files/tmp/etc/shadow_img_oscarimage already exists, exiting successfully.image:
$VAR1 = 'oscarimage';
---------------
Subcluster defined by image oscarimage wasn't validated!
... skipping it ...
Use of uninitialized value in numeric eq (==) at /opt/sync_files/bin/sync_files line 637.
Use of uninitialized value in numeric eq (==) at /opt/sync_files/bin/sync_files line 603.
[INFO](INFO]) File /opt/sync_files/tmp/etc/gshadow_debian already exists, exiting successfully.image:
$VAR1 = 'oscarimage';
---------------
Subcluster defined by image oscarimage wasn't validated!
... skipping it ...
Use of uninitialized value in numeric eq (==) at /opt/sync_files/bin/sync_files line 637.
Use of uninitialized value in numeric eq (==) at /opt/sync_files/bin/sync_files line 603.
ERROR: Impossible to update a file at /opt/sync_files/bin/sync_files line 467
        main::update_temp_file('/opt/sync_files/tmp/etc/passwd', '/opt/sync_files/tmp/etc/passwd', 'debian', 'oscarimage', 'passwd') called at /opt/sync_files/bin/sync_files line 516
        main::filter_file('/opt/sync_files/tmp/etc/passwd', '/opt/sync_files/tmp/etc/passwd', 'oscarimage', 'debian') called at /opt/sync_files/bin/sync_files line 620
        main::update_file('/opt/sync_files/tmp/etc/passwd', 'ARRAY(0x898fe70)') called at /opt/sync_files/bin/sync_files line 813
ERROR: Impossible to generate the temp file at /opt/sync_files/bin/sync_files line 517
        main::filter_file('/opt/sync_files/tmp/etc/passwd', '/opt/sync_files/tmp/etc/passwd', 'oscarimage', 'debian') called at /opt/sync_files/bin/sync_files line 620
        main::update_file('/opt/sync_files/tmp/etc/passwd', 'ARRAY(0x898fe70)') called at /opt/sync_files/bin/sync_files line 813
ERROR: Impossible to apply the file filter at /opt/sync_files/bin/sync_files line 621.

```

**Comment by dikim33** [comment]
**Comment by naughtont on 30 Jul 2009 21:31 UTC**
Current contents of the '/opt/sync_files/etc/sync_files.conf'

```
# This is the sync_files configuration file
# It must be located in /opt/sync_files/etc/sync_files.conf
# Values must be defined as shown
# Lines beginning with # are comments

# Uncomment to cause sync_files to ALWAYS 
# push ALL files (defaults to 0)
# force=1

# Defines where checksums are stored
# If the checksums have changed (or the file has not been created yet), 
# then it will be creatednext time sync_files is run.
# To change the checksum file location, uncomment the next line and change the
# file location
# checksums /opt/sync_files/etc/checksums

# Defines files to sync in a one file per line format, as below
# sync_files will check for existence before trying to sync them
syncfile /etc/passwd
syncfile /etc/group
syncfile /etc/shadow
syncfile /etc/gshadow
#syncfile /etc/sudoers
#syncfile /etc/modprobe.conf

#Using getent generates the following files in /opt/sync_files/tmp/, this is all 
#handled within sync_files, so they need not be listed here, this is only for
#clarification.
# syncfile /opt/sync_files/tmp/etc/passwd
# syncfile /opt/sync_files/tmp/etc/group
# syncfile /opt/sync_files/tmp/etc/shadow
# syncfile /opt/sync_files/tmp/etc/modprobe.conf

# 'getent' sets sync_files to use getent for the files listed above
# getent off

# sync_files uses C3 as its transport mechanism of choice.
# To turn off, uncomment the following line
# usec3 off

# sync_files also supports scp if set up to work with no passwords
# This can be used instead of c3, or in addition to it to sync files to hosts
# outside the default cluster defined in c3
# scphost node1
# scphost node3

# Defines where the crontab file is (usually /etc/crontab)
# Only change if necessary
# crontab="/etc/crontab"

# Defines how large the cron interval is (job runs every x minutes)
# This value is in minutes and must be less than 60
# If this is set to 0, then sync_files will not be set as a cron job
# To change the default value, uncomment the next line and change the value
# crondelay = 15

# Sometimes there needs to be manual modification to the crontab file
# In those cases, uncomment the following line.
# cronmodify = 0
```

**Comment by dikim33** [comment]
**Modified by valleegr on 29 Sep 2009 07:05 UTC**

**Comment by dikim33** [comment]
**Comment by valleegr on 6 Oct 2009 06:46 UTC**
r8844 fixes some of the problems and improve the output so it will be simpler to track the problem.

**Comment by dikim33** [comment]
**Comment by valleegr on 6 Oct 2009 07:26 UTC**
r8847 fixes even more problems. Now when running the command with the --force option, no errors are reported anymore but clearly, sc3, which is used to push files to compute nodes, does not do its work and the output is just useless.
Why do we use SC3 right now? Could we use C3 instead? That will simplify everything...

**Comment by dikim33** [comment]
**Comment by valleegr on 6 Oct 2009 21:08 UTC**
(In [8853]) This check-in fixes #565 - Last set of modifications to ensure files are synchronized on compute nodes.
